File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode prev = null;
        
        // Move the next node nth nodes from the beginning
        for (int i = 1; i < n; i++) {
            fast = fast.next;
        }
        
        // Move the fast to the end of the link
        while (fast.next != null) {
            prev = slow;
            fast = fast.next;
            slow = slow.next;
        }
        
        if (prev == null) {
            head = slow.next;
        } else {
            prev.next = prev.next.next;
            return head;
        }

        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev = dummy
        
        # Find nth node from the start
        p1 = p2 = head
        for _ in range(n - 1):
            p2 = p2.next
        
        # Find the nth node from the end
        while p2.next:
            prev = p1
            p1 = p1.next
            p2 = p2.next
        
        prev.next = prev.next.next # remove the nth node from the end
        return dummy.next